src/main/java/com/openlattice/shuttle/dates/JavaDateTimeHelper.java

```diff
@@ -1,88 +1,94 @@
                 }
+            } catch ( ExecutionException ex ) {
+                logger.error("ExecutionException loading pattern from cache", ex);
             }
         }
         return null;
     }
 
-    public OffsetDateTime parseDateTime(String date) {
-        if (shouldIgnoreValue(date))
-            return null;
-        for (int i = 0; i < datePatterns.length; ++i) {
-            DateTimeFormatter formatter = formatters.get(i);
-            try {
-                LocalDateTime ldt = LocalDateTime.parse(date, formatter);
-                //                if (ldt.isBefore(LocalDateTime.of(1900, 1, 1, 0, 0, 0))){return null;}
-                if (
-                        datePatterns[0].matches(".*yy.*") && !datePatterns[0].matches(".*yyyy.*") ||
-                                datePatterns[0].matches(".*YY.*") && !datePatterns[0].matches(".*YYYY.*")
-                ) {
-                    if ((ldt.getYear() - LocalDate.now().getYear()) > 20) {
-                        ldt = ldt.withYear(ldt.getYear() - 100);
-                    }
+    public LocalDate parseDate(String date) {
+        return parseWithTwoDigitYearHandling( date, LocalDate::parse, ( ld, datePattern ) -> {
+            if (  checkDatePatternIsTwoDigitYear( datePattern ) ) {
+                // TODO: break this out into its own transform that specifies the date boundaries for two-digit years
+                if ((ld.getYear() - LocalDate.now().getYear()) > 20) {
+                    ld = ld.withYear(ld.getYear() - 100);
                 }
-                return ldt.atZone(tz.toZoneId()).toOffsetDateTime();
-            } catch (Exception e) {
-                if (i == datePatterns.length - 1) {
-                    logger.error("Unable to parse datetime {}, please see debug log for additional information: {}.",
-                            date, e);
+            }
+            return ld;
+        });
+    }
+
+    public OffsetDateTime parseDateTime(String date) {
+        return parseWithTwoDigitYearHandling( date, ( toParse, formatter ) -> {
+            LocalDateTime parsed = LocalDateTime.parse( toParse, formatter );
+            return parsed.atZone( tz.toZoneId() ).toOffsetDateTime();
+        }, ( odt, datePattern ) -> {
+            if (  checkDatePatternIsTwoDigitYear( datePattern ) ) {
+                // TODO: break this out into its own transform that specifies the date boundaries for two-digit years
+                if ((odt.getYear() - LocalDate.now().getYear()) > 20) {
+                    odt = odt.withYear(odt.getYear() - 100);
                 }
             }
-        }
-        return null;
+            return odt;
+        });
+    }
+
+    private boolean checkDatePatternIsTwoDigitYear( String datePattern ) {
+        return !datePattern.endsWith( "yyyy" ) && !datePattern.startsWith("yyyy") && ( datePattern.endsWith( "yy" ) || datePattern.startsWith( "yy" ) );
```

Comment:
I don't think it's guaranteed that the date pattern will either start or end with the year (especially for datetimes) -- also are the patterns all being lowercased?

Reply:
I grep'd through all of our flight.yamls and it looks like we never use `YY` or `YYYY` and `yy` and `YY` have slightly different meanings anyway and I believe we shouldn't be using `YY/YYYY` anywhere anyway.

Reply:
Oh and yes, that's correct that there are no guarantees on starting/ending with `yy`/`yyyy` but across all of our `flight.yaml` that's the current pattern.
I can change this to a better regex if you think I should, but anyone/where who writes a date 01/22/03 and the 22 is the year should probably not be allowed near a computer. 
https://www.juandebravo.com/2015/04/10/java-yyyy-date-format/I think for `yy` style dates going forward we should have a different transform where the boundaries of the date are specified ie. `dd/mm/yy, 1916, present` would mean that there are conflicting full year values for `yy` == 16, 17, 18, 19, etc.

Reply:
Is this function only used for date patterns (and not datetime patterns)?

Reply:
and yeah Y is week-based year so if someone uses that it's probably a bug -- so maybe on registering date(time) patterns we should throw an exception (or at least log something) if someone uses a pattern with a Y

Reply:
^would probably be a really subtle bug too, since most dates would be parsed correctly

Reply:
No, it's used for both and used indirectly to add `yy` format support all over the place. Though we only have dates that use `yy` in two integrations (as far as I can tell).
```
DateAsDateTimeTransform
DateTransform
DateTimeAsDateTransform
CombineDateTimeTransform
DateTimeTransform
```

Reply:
Re: bug
https://www.juandebravo.com/2015/04/10/java-yyyy-date-format/

Reply:
yeah, that's why I feel like it should be treated as a bug

Reply:
And if this is being used on datetimes, then I think we should expect to come across integrations with datetime patterns where years that aren't at the beginning or end

Reply:
Ohhh right, forgot about the format options there. Yeah I'll update it for that then

Reply:
Updated to handle `yy` as previous, just need to add exception cases for `YY`/`YYYY`